src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -15,661 +15,693 @@
+         typename InitializationRuleType,
+         typename MatType>
+void FFN<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Reset(const size_t inputDimensionality)
 {
-  boost::apply_visitor(ForwardVisitor(inputs,
-      boost::apply_visitor(outputParameterVisitor, network[begin])),
-      network[begin]);
+  parameters.clear();
 
-  for (size_t i = 1; i < end - begin + 1; ++i)
+  // If the user provided an input dimensionality, then we will take that as the
+  // new input size.  Otherwise, whatever is currently specified in
+  // `InputDimensions()` will be used.
+  if (inputDimensionality != 0)
+  {
+    CheckNetwork("FFN::Reset()", inputDimensionality, true, false);
+  }
+  else
   {
-    boost::apply_visitor(ForwardVisitor(boost::apply_visitor(
-        outputParameterVisitor, network[begin + i - 1]),
-        boost::apply_visitor(outputParameterVisitor, network[begin + i])),
-        network[begin + i]);
+    const size_t inputDims = std::accumulate(inputDimensions.begin(),
+        inputDimensions.end(), 0);
+    CheckNetwork("FFN::Reset()", inputDims, true, false);
   }
+}
 
-  results = boost::apply_visitor(outputParameterVisitor, network[end]);
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void FFN<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::SetNetworkMode(const bool training)
+{
+  network.Training() = training;
 }
 
-template<typename OutputLayerType, typename InitializationRuleType,
-         typename... CustomLayers>
-template<typename PredictorsType, typename TargetsType, typename GradientsType>
-double FFN<OutputLayerType, InitializationRuleType, CustomLayers...>::Backward(
-    const PredictorsType& inputs,
-    const TargetsType& targets,
-    GradientsType& gradients)
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void FFN<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Forward(const MatType& inputs, MatType& results)
 {
-  double res = outputLayer.Forward(boost::apply_visitor(
-      outputParameterVisitor, network.back()), targets);
+  Forward(inputs, results, 0, network.Network().size() - 1);
+}
 
-  for (size_t i = 0; i < network.size(); ++i)
-  {
-    res += boost::apply_visitor(lossVisitor, network[i]);
-  }
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void FFN<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Forward(const MatType& inputs,
+           MatType& results,
+           const size_t begin,
+           const size_t end)
+{
+  // Sanity checking...
+  if (end < begin)
+    return;
+
+  // Ensure the network is valid.
+  CheckNetwork("FFN::Forward()", inputs.n_rows);
+
+  // We must always store a copy of the forward pass in `networkOutputs` in case
+  // we do a backward pass.
+  networkOutput.set_size(network.OutputSize(), inputs.n_cols);
```

Comment:
If the users uses `Predict` we already know we don't need the outputs for each layer, so we could avoid an extra copy here?

Reply:
Nice point!  I changed `Predict()` to use `network.Forward()` instead of the existing `FFN::Forward()` implementation, which means that it won't make a copy of the output.  The patch was way easier than I thought: 07be731.